File: bionic-js-java/src/test/java/acceptance/toy1/ToyClass2Test.java
package acceptance.toy1;

import acceptance.toy1.bjs.BjsTestProject;
import acceptance.toy1.bjs.ToyClass2;
import bionic.js.BjsAnyObject;
import bionic.js.Lambda;
import jjbridge.api.value.JSType;
import jjbridge.engine.v8.V8Engine;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.Date;

import static org.junit.jupiter.api.Assertions.*;

public class ToyClass2Test
{
    private Date testDate;
    private Double testDateTimestamp;
    private ToyClass2 toyObj;
    private ToyClass2 toy;

    @BeforeAll
    public static void beforeClass()
    {
        BjsTestProject.setJsEngine(new V8Engine());
    }

    @BeforeEach
    public void before()
    {
        testDateTimestamp = 472867200000D;
        testDate = Date.from(Instant.ofEpochMilli(testDateTimestamp.longValue()));
        toyObj = new ToyClass2(false);
        toyObj.log("any object");
        
        toy = new ToyClass2(false);
        assertEquals("called constructor without params", toy.log());
    }

    @Test
    public void testConstructorParam() {
        toy = new ToyClass2(true);
        assertEquals("called constructor with params: true", toy.log());
    }

    @Test
    public void testConstructorNullParam() {
        toy = new ToyClass2((Boolean) null);
        assertEquals("called constructor without params", toy.log());
    }

    @Test
    public void testNativeAutoProp()
    {
        assertNull(toy.nativeAutoProp());

        toy.nativeAutoProp("1984!");
        assertEquals("1984!", toy.nativeAutoProp());

        toy.nativeAutoProp(null);
        assertNull(toy.nativeAutoProp());
    }

    @Test
    public void testAnyAutoProp() {
        assertEquals(toy.anyAutoProp().jsObj.getActualType(), JSType.Undefined);

        toy.anyAutoProp(new BjsAnyObject(toyObj));
        assertEquals("any object", toy.anyAutoProp().getObject(ToyClass2.bjsFactory, ToyClass2.class).log());
    }

    @Test
    public void testBjsObjAutoProp() {
        assertNull(toy.bjsObjAutoProp());

        toy.bjsObjAutoProp(toyObj);
        assertEquals("any object", toy.bjsObjAutoProp().log());
    }

    @Test
    public void testLambdaAutoProp() {
        assertNull(toy.lambdaAutoProp());

        toy.lambdaAutoProp(string -> string == null ? "null!" : string + " 1984!");
        assertEquals("null!", toy.lambdaAutoProp().apply(null));
        assertEquals("W 1984!", toy.lambdaAutoProp().apply("W"));

        toy.lambdaAutoProp(null);
        assertNull(toy.lambdaAutoProp());

        toy.lambdaAutoProp(string -> null);
        assertNull(toy.lambdaAutoProp().apply("x"));
    }

    @Test
    public void testLambdaAutoProp_putNullToJs() {
        String nullTest = "(this.lambdaAutoProp() === null)";

        toy.lambdaAutoProp(string -> "hey");
        toy.evalAndSetValue(nullTest);
        assertFalse(toy.boolFunc());

        toy.lambdaAutoProp(string -> null);
        toy.evalAndSetValue(nullTest);
        assertTrue(toy.boolFunc());
    }

    @Test
    public void testNativeArrayAutoProp() {
        assertNull(toy.nativeArrayAutoProp());

        toy.nativeArrayAutoProp(new String[][][]{{{"a", "b"}}});
        assertArrayEquals(new String[][][]{{{"a", "b"}}}, toy.nativeArrayAutoProp());

        toy.nativeArrayAutoProp(new String[][][]{{{}, {}}});
        assertArrayEquals(new String[][][]{{{}, {}}}, toy.nativeArrayAutoProp());

        toy.nativeArrayAutoProp(new String[][][]{{}, {}});
        assertArrayEquals(new String[][][]{{}, {}}, toy.nativeArrayAutoProp());

        toy.nativeArrayAutoProp(new String[][][]{});
        assertArrayEquals(new String[][][]{}, toy.nativeArrayAutoProp());

        toy.nativeArrayAutoProp(new String[][][]{null, {null}, {{null}}, {{"test", null}, null}});
        assertArrayEquals(new String[][][]{null, {null}, {{null}}, {{"test", null}, null}}, toy.nativeArrayAutoProp());
    }

    @Test
    public void testBjsObjArrayAutoProp() {
        assertNull(toy.bjsObjArrayAutoProp());

        toy.bjsObjArrayAutoProp(new ToyClass2[][][]{{{toyObj}}});
        assertArrayEquals(new ToyClass2[][][]{{{toyObj}}}, toy.bjsObjArrayAutoProp());

        toy.bjsObjArrayAutoProp(new ToyClass2[][][]{{{}, {}}});
        assertArrayEquals(new ToyClass2[][][]{{{}, {}}}, toy.bjsObjArrayAutoProp());

        toy.bjsObjArrayAutoProp(new ToyClass2[][][]{{}, {}});
        assertArrayEquals(new ToyClass2[][][]{{}, {}}, toy.bjsObjArrayAutoProp());

        toy.bjsObjArrayAutoProp(new ToyClass2[][][]{});
        assertArrayEquals(new ToyClass2[][][]{}, toy.bjsObjArrayAutoProp());

        toy.bjsObjArrayAutoProp(new ToyClass2[][][]{null, {null}, {{null}}, {{toyObj, null}, null}});
        assertArrayEquals(new ToyClass2[][][]{null, {null}, {{null}}, {{toyObj, null}, null}}, toy.bjsObjArrayAutoProp());
    }

    @Test
    public void testAnyArrayAutoProp() {
        assertNull(toy.anyArrayAutoProp());

        BjsAnyObject anyObj = new BjsAnyObject(toyObj);

        toy.anyArrayAutoProp(new BjsAnyObject[][][]{{{anyObj}}});
        assertEquals("any object", toy.anyArrayAutoProp()[0][0][0].getObject(ToyClass2.bjsFactory, ToyClass2.class).log());

        toy.anyArrayAutoProp(new BjsAnyObject[][][]{{{}}});
        assertEquals(0, toy.anyArrayAutoProp()[0][0].length);
    }

    @Test
    public void testLambdaArrayAutoProp() {
        assertNull(toy.lambdaArrayAutoProp());

        Lambda.F1<String, String> lambda1a = s -> (s != null ? s : "") + "-1a";
        Lambda.F1<String, String> lambda1b = s -> (s != null ? s : "") + "-1b";
        Lambda.F1<String, String> lambda2a = s -> (s != null ? s : "") + "-2a";

        toy.lambdaArrayAutoProp(new Lambda.F1[][][]{{{lambda1a, lambda1b}, {lambda2a, null}}});

        assertEquals("test-1a", toy.lambdaArrayAutoProp()[0][0][0].apply("test"));
        assertEquals("test-1b", toy.lambdaArrayAutoProp()[0][0][1].apply("test"));
        assertEquals("test-2a", toy.lambdaArrayAutoProp()[0][1][0].apply("test"));
        assertNull(toy.lambdaArrayAutoProp()[0][1][1]);

        toy.lambdaArrayAutoProp(new Lambda.F1[][][]{{{}}});
        assertEquals(0, toy.lambdaArrayAutoProp()[0][0].length);
    }

    @Test
    public void testProp() {
        assertEquals("1984!", toy.prop());

        toy.prop("test value");
        assertEquals("test value", toy.prop());

        toy.prop(null);
        assertEquals("1984!", toy.prop());
    }

    @Test
    public void testVoidFunc() {
        toy.voidFunc();
        assertEquals("called voidFunc", toy.log());
    }

    @Test
    public void testParamsFunc() {
        toy.paramsFunc(true, testDate, 01.984, 1984, "1984", new BjsAnyObject(toyObj), toyObj,
                new Integer[]{1,2,3}, () -> "lambda return value");
        assertEquals("called paramsFunc with params: true, 1984-12-26T00:00:00.000Z, 1.984, 1984, 1984, any object, any object, [1,2,3], lambda return value", toy.log());
    }

    @Test
    public void testParamsFunc_null() {
        toy.paramsFunc(null, null, null, null, null, ToyClass2.bjs.anyNull(), null, null, null);
        assertEquals("called paramsFunc with params: null, null, null, null, null, null, null, null, null", toy.log());
    }

    @Test
    public void testBoolFunc() {
        assertNull(toy.boolFunc());
        assertEquals("called retValueFunc", toy.log());

        toy.evalAndSetValue("null");
        assertNull(toy.boolFunc());

        toy.evalAndSetValue("false");
        assertFalse(toy.boolFunc());

        toy.evalAndSetValue("true");
        assertTrue(toy.boolFunc());

        toy.evalAndSetValue("!!1");
        assertTrue(toy.boolFunc());
    }

    @Test
    public void testDateFunc() {
        assertNull(toy.dateFunc());
        assertEquals("called retValueFunc", toy.log());

        toy.evalAndSetValue("null");
        assertNull(toy.dateFunc());

        toy.evalAndSetValue("new Date(" + testDateTimestamp + ")");
        assertEquals(testDate, toy.dateFunc());
    }

    @Test
    public void testFloatFunc() {
        assertNull(toy.floatFunc());
        assertEquals("called retValueFunc", toy.log());

        toy.evalAndSetValue("null");
        assertNull(toy.floatFunc());

        toy.evalAndSetValue(".1984");
        assertEquals(0.1984, toy.floatFunc(), 0.0);

        toy.evalAndSetValue("123456789.123456789");
        assertEquals(123456789.123456789, toy.floatFunc(), 0.0);
    }

    @Test
    public void testIntFunc() {
        assertNull(toy.intFunc());
        assertEquals("called retValueFunc", toy.log());

        toy.evalAndSetValue("null");
        assertNull(toy.intFunc());

        toy.evalAndSetValue("1984");
        assertEquals(1984, (long) toy.intFunc());

        toy.evalAndSetValue("1234567890");
        assertEquals(1234567890, (long) toy.intFunc());
    }

    @Test
    public void testStringFunc() {
        assertNull(toy.stringFunc());
        assertEquals("called retValueFunc", toy.log());

        toy.evalAndSetValue("null");
        assertNull(toy.stringFunc());

        toy.evalAndSetValue("'hello 84! 👋'");
        assertEquals("hello 84! 👋", toy.stringFunc());
    }

    @Test
    public void testLambdaVoidFunc() {
        Boolean[] check = new Boolean[] {false};

        Lambda.F0<Void> jsLambda = toy.lambdaVoidFunc(() -> {
           assertEquals("calling lambda", toy.log());
           check[0] = true;
           return null;
        });

        assertEquals("called lambdaVoidFunc", toy.log());
        jsLambda.apply();
        assertTrue(check[0]);

        Lambda.F0<Void> nilLambda = toy.lambdaVoidFunc(null);
        assertNull(nilLambda);
    }

    @Test
    public void testLambdaWithParamsFunc() {
        toy.lambdaWithParamsFunc((integer, nullStr, any, obj) -> {
            assertEquals("called lambdaWithParamsFunc", toy.log());
            assertEquals(1984, (int) integer);
            assertNull(nullStr);
            assertEquals("hello 84", any.getObject(ToyClass2.bjsFactory, ToyClass2.class).log());
            assertEquals("hello 84", obj.log());
            return "1984!";
        });
        assertEquals("called lambda with result: 1984!", toy.log());
    }

    @Test
    public void testReturningLambdaWithParamsFunc() {
        Lambda.F4<Integer, BjsAnyObject, ToyClass2, Integer[], String> lambda = toy.returningLambdaWithParamsFunc();
        assertEquals("called returningLambdaWithParamsFunc", toy.log());
        String lambdaRetValue = lambda.apply(1984, new BjsAnyObject(toyObj), toyObj, new Integer[]{1,2,3});
        assertEquals("called returned lambda with params: 1984, any object, any object, [1,2,3]", toy.log());
        assertEquals("lambda returning value", lambdaRetValue);
    }

    @Test
    public void testReturningLambdaWithParamsFunc_null() {
        Lambda.F4<Integer, BjsAnyObject, ToyClass2, Integer[], String> lambda = toy.returningLambdaWithParamsFunc();
        assertEquals("called returningLambdaWithParamsFunc", toy.log());
        String lambdaRetValue = lambda.apply(null, ToyClass2.bjs.anyNull(), null, null);
        assertEquals("called returned lambda with params: null, null, null, null", toy.log());
        assertEquals("lambda returning value", lambdaRetValue);
    }
}

File: bionic-js-swift/BjsTests/Tests/ToyClass2Test.swift
import XCTest
import JavaScriptCore

class ToyClass2Test: XCTestCase {
    
    var testDate: Date!
    var testDateTimestamp: Double!
    var toyObj: ToyClass2!
    var toy: ToyClass2!
    
    override func setUp() {
        super.setUp()
        testDateTimestamp = 472867200
        testDate = Date(timeIntervalSince1970: testDateTimestamp)
        toyObj = ToyClass2(false)
        toyObj.log = "any object"
        
        toy = ToyClass2(false)
        XCTAssertEqual(toy.log, "called constructor without params")
    }
    
    
    // CONSTRUCTOR
    
    func testConstructorParam() {
        let toy = ToyClass2(true)
        XCTAssertEqual(toy.log, "called constructor with params: true")
    }
    
    func testConstructorNullParam() {
        let toy = ToyClass2(nil)
        XCTAssertEqual(toy.log, "called constructor without params")
    }
    
    
    // INSTANCE PROPERTIES
    
    func testNativeAutoProp() {
        XCTAssertNil(toy.nativeAutoProp)
        
        toy.nativeAutoProp = "1984!"
        XCTAssertEqual(toy.nativeAutoProp, "1984!")
        
        toy.nativeAutoProp = nil
        XCTAssertNil(toy.nativeAutoProp)
    }
    
    func testAnyAutoProp() {
        XCTAssertTrue(toy.anyAutoProp.jsObj.isUndefined)
        
        toy.anyAutoProp = BjsAnyObject(toyObj)
        XCTAssertEqual(toy.anyAutoProp.getObject(ToyClass2.bjsFactory)!.log, "any object")
    }
    
    func testBjsObjAutoProp() {
        XCTAssertNil(toy.bjsObjAutoProp)
        
        toy.bjsObjAutoProp = toyObj
        XCTAssertEqual(toy.bjsObjAutoProp!.log, "any object")
    }
    
    func testLambdaAutoProp() {
        XCTAssertNil(toy.lambdaAutoProp)
        toy.lambdaAutoProp = { string in
            if string == nil {
                return "null!"
            }
            return string! + " 1984!"
        }
        XCTAssertEqual(toy.lambdaAutoProp!(nil), "null!")
        XCTAssertEqual(toy.lambdaAutoProp!("W"), "W 1984!")
        
        toy.lambdaAutoProp = nil
        XCTAssertNil(toy.lambdaAutoProp)
        
        toy.lambdaAutoProp = { string in
            return nil
        }
        XCTAssertEqual(toy.lambdaAutoProp!("x"), nil)
    }
    
    func testLambdaAutoProp_putNilToJs() {
        let nullTest = "(this.lambdaAutoProp() === null)"
        
        toy.lambdaAutoProp = { string in return "hey" }
        toy.evalAndSetValue(nullTest)
        XCTAssertEqual(toy.boolFunc(), false)
        
        toy.lambdaAutoProp = { string in return nil }
        toy.evalAndSetValue(nullTest)
        XCTAssertEqual(toy.boolFunc(), true)
    }
    
    func testNativeArrayAutoProp() {
        XCTAssertNil(toy.nativeArrayAutoProp)
        
        toy.nativeArrayAutoProp = [[["a", "b"]]]
        XCTAssertEqual(toy.nativeArrayAutoProp, [[["a", "b"]]])
        
        toy.nativeArrayAutoProp = [[[], []]]
        XCTAssertEqual(toy.nativeArrayAutoProp, [[[], []]])
        
        toy.nativeArrayAutoProp = [[], []]
        XCTAssertEqual(toy.nativeArrayAutoProp, [[], []])
        
        toy.nativeArrayAutoProp = []
        XCTAssertEqual(toy.nativeArrayAutoProp, [])
        
        toy.nativeArrayAutoProp = [nil, [nil], [[nil]], [["test", nil]]]
        XCTAssertEqual(toy.nativeArrayAutoProp, [nil, [nil], [[nil]], [["test", nil]]])
    }
    
    func testBjsObjArrayAutoProp() {
        XCTAssertNil(toy.bjsObjArrayAutoProp)
        
        toy.bjsObjArrayAutoProp = [[[toyObj]]]
        XCTAssertEqual(toy.bjsObjArrayAutoProp, [[[toyObj]]])
        
        toy.bjsObjArrayAutoProp = [[[], []]]
        XCTAssertEqual(toy.bjsObjArrayAutoProp, [[[], []]])
        
        toy.bjsObjArrayAutoProp = [[], []]
        XCTAssertEqual(toy.bjsObjArrayAutoProp, [[], []])
        
        toy.bjsObjArrayAutoProp = []
        XCTAssertEqual(toy.bjsObjArrayAutoProp, [])
        
        toy.bjsObjArrayAutoProp = [nil, [nil], [[nil]], [[toyObj, nil]]]
        XCTAssertEqual(toy.bjsObjArrayAutoProp, [nil, [nil], [[nil]], [[toyObj, nil]]])
    }
    
    func testAnyArrayAutoProp() {
        XCTAssertNil(toy.anyArrayAutoProp)
        
        let anyObj = BjsAnyObject(toyObj.bjsObj)
        
        toy.anyArrayAutoProp = [[[ anyObj ]]]
        XCTAssertEqual(toy.anyArrayAutoProp![0]![0]![0].getObject(ToyClass2.bjsFactory)?.log, "any object")
        
        toy.anyArrayAutoProp = [[[ ]]]
        XCTAssertTrue(toy.anyArrayAutoProp![0]![0]!.isEmpty)
    }
    
    func testLambdaArrayAutoProp() {
        XCTAssertNil(toy.lambdaArrayAutoProp)
        
        let lambda1a: (_ inStr: String?) -> String? = { ($0 ?? "") + "-1a"}
        let lambda1b: (_ inStr: String?) -> String? = { ($0 ?? "") + "-2a"}
        let lambda2a: (_ inStr: String?) -> String? = { ($0 ?? "") + "-1b"}
        
        toy.lambdaArrayAutoProp = [[[ lambda1a, lambda1b ], [lambda2a, nil] ]]
        
        XCTAssertEqual(toy.lambdaArrayAutoProp![0]![0]![0]!("test"), "test-1a")
        XCTAssertEqual(toy.lambdaArrayAutoProp![0]![0]![1]!("test"), "test-2a")
        XCTAssertEqual(toy.lambdaArrayAutoProp![0]![1]![0]!("test"), "test-1b")
        XCTAssertNil(toy.lambdaArrayAutoProp![0]![1]![1])
        
        toy.lambdaArrayAutoProp = [[[ ]]]
        XCTAssertTrue(toy.lambdaArrayAutoProp![0]![0]!.isEmpty)
    }
    
    func testProp() {
        XCTAssertEqual(toy.prop, "1984!")
        
        toy.prop = "test value"
        XCTAssertEqual(toy.prop, "test value")
        
        toy.prop = nil
        XCTAssertEqual(toy.prop, "1984!")
    }
    
    
    // INSTANCE METHODS
    
    func testVoidFunc() {
        toy.voidFunc()
        XCTAssertEqual(toy.log, "called voidFunc")
    }
    
    func testParamsFunc() {
        let testExpectation = expectation(description: "called")
        toy.paramsFunc(true, testDate, 01.984, 01984, "1984", BjsAnyObject(toyObj.bjsObj), toyObj,
                       [1,2,3]) { () -> String? in
            testExpectation.fulfill()
            return "lambda return value"
        }
        XCTAssertEqual(toy.log, "called paramsFunc with params: true, 1984-12-26T00:00:00.000Z, 1.984, 1984, 1984, any object, any object, [1,2,3], lambda return value")
        wait(for: [testExpectation], timeout: 10)
    }
    
    func testParamsFunc_nil() {
        toy.paramsFunc(nil, nil, nil, nil, nil, ToyClass2.bjs.anyNull, nil, nil, nil)
        XCTAssertEqual(toy.log, "called paramsFunc with params: null, null, null, null, null, null, null, null, null")
    }
    
    func testBoolFunc() {
        XCTAssertNil(toy.boolFunc())
        XCTAssertEqual(toy.log, "called retValueFunc")
        
        toy.evalAndSetValue("null")
        XCTAssertNil(toy.boolFunc())
        
        toy.evalAndSetValue("false")
        XCTAssertEqual(toy.boolFunc(), false)
        
        toy.evalAndSetValue("true")
        XCTAssertEqual(toy.boolFunc(), true)
        
        toy.evalAndSetValue("1")
        XCTAssertEqual(toy.boolFunc(), true)
    }
    
    func testDateFunc() {
        XCTAssertNil(toy.dateFunc())
        XCTAssertEqual(toy.log, "called retValueFunc")
        
        toy.evalAndSetValue("null")
        XCTAssertNil(toy.dateFunc())
        
        toy.evalAndSetValue("new Date(\(testDateTimestamp!) * 1000)")
        XCTAssertEqual(toy.dateFunc(), testDate)
    }
    
    func testFloatFunc() {
        XCTAssertNil(toy.floatFunc())
        XCTAssertEqual(toy.log, "called retValueFunc")
        
        toy.evalAndSetValue("null")
        XCTAssertNil(toy.floatFunc())
        
        toy.evalAndSetValue(".1984")
        XCTAssertEqual(toy.floatFunc(), 0.1984)
        
        toy.evalAndSetValue("123456789.123456789")
        XCTAssertEqual(toy.floatFunc(), 123456789.123456789)
    }
    
    func testIntFunc() {
        XCTAssertNil(toy.intFunc())
        XCTAssertEqual(toy.log, "called retValueFunc")
        
        toy.evalAndSetValue("null")
        XCTAssertNil(toy.intFunc())
        
        toy.evalAndSetValue("1984")
        XCTAssertEqual(toy.intFunc(), 1984)
        
        toy.evalAndSetValue("1234567890")
        XCTAssertEqual(toy.intFunc(), 1234567890)
    }
    
    func testStringFunc() {
        XCTAssertNil(toy.stringFunc())
        XCTAssertEqual(toy.log, "called retValueFunc")
        
        toy.evalAndSetValue("null")
        XCTAssertNil(toy.stringFunc())
        
        toy.evalAndSetValue("'hello 84! 👋'")
        XCTAssertEqual(toy.stringFunc(), "hello 84! 👋")
    }
    
    func testLambdaVoidFunc() {
        let testExpectation = expectation(description: "called")
        
        let jsLambda = toy.lambdaVoidFunc {
            XCTAssertEqual(self.toy.log, "calling lambda")
            testExpectation.fulfill()
        }
        XCTAssertEqual(toy.log, "called lambdaVoidFunc")
        jsLambda!()
        wait(for: [testExpectation], timeout: 10)
        
        let nilLambda = toy.lambdaVoidFunc(nil)
        XCTAssertNil(nilLambda)
    }
    
    func testLambdaVoidFunc_nil() {
        let nilLambda = toy.lambdaVoidFunc(nil)
        XCTAssertNil(nilLambda)
    }
    
    func testLambdaWithParamsFunc() {
        let testExpectation = expectation(description: "called")
        toy.lambdaWithParamsFunc { (int, nullStr, any, obj) in
            XCTAssertEqual(self.toy.log, "called lambdaWithParamsFunc")
            XCTAssertEqual(int, 1984)
            XCTAssertNil(nullStr)
            XCTAssertEqual(any.getObject(ToyClass2.bjsFactory)?.log, "hello 84")
            XCTAssertEqual(obj?.log, "hello 84")
            testExpectation.fulfill()
            return "1984!"
        }
        XCTAssertEqual(toy.log, "called lambda with result: 1984!")
        wait(for: [testExpectation], timeout: 10)
    }
    
    func testReturningLambdaWithParamsFunc() {
        let lambda = toy.returningLambdaWithParamsFunc()
        XCTAssertEqual(toy.log, "called returningLambdaWithParamsFunc")
        let lambdaRetValue = lambda!(1984, BjsAnyObject(toyObj.bjsObj), toyObj, [1,2,3])
        XCTAssertEqual(toy.log, "called returned lambda with params: 1984, any object, any object, [1,2,3]")
        XCTAssertEqual(lambdaRetValue, "lambda returning value")
    }
    
    func testReturningLambdaWithParamsFunc_nil() {
        let lambda = toy.returningLambdaWithParamsFunc()
        XCTAssertEqual(toy.log, "called returningLambdaWithParamsFunc")
        let lambdaRetValue = lambda!(nil, ToyClass2.bjs.anyNull, nil, nil)
        XCTAssertEqual(toy.log, "called returned lambda with params: null, null, null, null")
        XCTAssertEqual(lambdaRetValue, "lambda returning value")
    }
}
